### PR TITLE
feat(editor): allow toggling between old/new editors

### PR DESCRIPTION
**Pull request by seaerchin**
**Next PR is #1624** 

## Problem
We want to give users a way to toggle their pages between the old and new RTE so that they don't lose data. 

This PR adds a toggle in the page creation modal that explicitly allows the users to do this conversion. Do take note this is **not** the final toggle location - this will be decided by product/design. 

## Solution
1. Add a key in our frontmatter, `variant` that details if it is `tiptap` or `markdown` editor. This will determine the _default_ state of the toggle. We show the old RTE depending on the toggle state. 
2. Rework the main `EditPage` - this handles toggling between the "old" and the "new" editor. This now handles toggling of state, together with providing modal/editor contexts down to the children.
    - we require the new tiptap editor (as in the state manager) **even in the markdown** as we want to show a preview modal that has the new RTE. 
    - this editor state is reset when rejecting the update -> we only use the fetched state from github if they are originally on tiptap (from frontmatter) + the current state is `tiptap`. This is to ensure that it's the first load of the new editor.
3. Add a new `MarkdownEditPage` - this is **mostly** copied over, together with new stuff
    - an alert at the top of the page that shows a preview modal that does the toggling
    - **TAKE NOTE:** additional functionality for displaying the original modals was removed -> will file ticket to add back in. wanted simplest version of old editor.

## Tests 
**Markdown RTE**
- [ ] create a new page using the old RTE
- [ ] type some content in the page (version A) 
- [ ] click preview 
- [ ] type some content inside the editor (version A')
- [ ] click close (ie, reject update)
- [ ] content should not change (still show version A in old RTE) 
- [ ] type some extra content in old RTE (version B) 
- [ ] click preview 
- [ ] the content inside the new RTE should show version B.
- [ ] type some content in new RTE (B') 
- [ ] click `I'm in`
- [ ] new RTE shuold display with content B'

**Tiptap RTE**
- [ ] create a new page using the new RTE
- [ ] click into page 
- [ ] should show new RTE
- [ ] make some changes
- [ ] click save
- [ ] exit the page
- [ ] click back in
- [ ] changes shuold still be there
- [ ] should still use **new** RTE 
- [ ] exit the page
- [ ] toggle to old RTE
- [ ] go back in
- [ ] should display content as html

### CODE REVIEW

File: src/layouts/components/Editor/components/TableBubbleMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          action={() => editor.chain().focus().addRowAfter().run()}
+          title="Add row after"
+          isRound
+        />
+        <MenuItem
+          icon="delete-row"
+          action={() => editor.chain().focus().deleteRow().run()}
+          title="Delete row"
+          isRound
+        />
```
Comment (seaerchin) [inline src/layouts/components/Editor/components/TableBubbleMenu.tsx:61]: new changes, rest are copy pasted

File: src/layouts/components/Editor/extensions/Iframe/IframeView.tsx
Comment (seaerchin) [inline src/layouts/components/Editor/extensions/Iframe/IframeView.tsx:1]: no new changes, just renamed

File: src/layouts/EditPage/TiptapEditPage.tsx
Comment (seaerchin) [inline src/layouts/EditPage/TiptapEditPage.tsx:1]: no new changes, sharded out from existing file

File: src/layouts/EditPage/EditPageLayout.tsx
Comment (seaerchin) [inline src/layouts/EditPage/EditPageLayout.tsx:1]: just shifted to new file

File: src/layouts/EditPage/MarkdownEditPage.tsx [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    updateMediaSrcs()
+  }, [csp, editorValue])
+
+  useEffect(() => {
+    updateHtmlWithMediaData()
+  }, [mediaData, editorValue])
+
+  return (
+    <EditPageLayout getPageBody={() => editorValue} variant="markdown">
+      <Box maxW="50%" p="1.25rem">
```
Comment (seaerchin) [inline src/layouts/EditPage/MarkdownEditPage.tsx:160]: new change

File: src/layouts/EditPage/MarkdownEditPage.tsx [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    </EditPageLayout>
+  )
+}
+
+interface PreviewModalProps {
+  isOpen: boolean
+  onClose: () => void
+  togglePreview: () => void
+}
+const PreviewModal = ({
```
Comment (seaerchin) [inline src/layouts/EditPage/MarkdownEditPage.tsx:215]: new change

File: src/layouts/EditPage/TiptapEditPage.tsx
Comment (seaerchin) [inline src/layouts/EditPage/TiptapEditPage.tsx:1]: shifted out, no changes

File: src/layouts/components/Editor/extensions/Iframe/IframeView.tsx
Comment (seaerchin) [inline src/layouts/components/Editor/extensions/Iframe/IframeView.tsx:1]: No new changes, just renamed

File: src/layouts/EditPage/TiptapEditPage.tsx
Comment (seaerchin) [inline src/layouts/EditPage/TiptapEditPage.tsx:1]: No new changes, sharded from existing file

File: src/layouts/EditPage/EditPageLayout.tsx
Comment (seaerchin) [inline src/layouts/EditPage/EditPageLayout.tsx:1]: shifted to new file

**Review by kishore03109 — APPROVED** [review]
I dont have blocking comments but I am worried about rollout. 
1. is this feature flagged + effect on non-feature flagged users
2. UI polish if all users are going to see that loading blip
3. do we have a known list of bugs that we know but we intend to fix them post beta testing?

**Comment by seaerchin** [comment]
> I dont have blocking comments but I am worried about rollout.
> 
> 1. is this feature flagged + effect on non-feature flagged users
> 2. UI polish if all users are going to see that loading blip
> 3. do we have a known list of bugs that we know but we intend to fix them post beta testing?

this is flagged. non flagged users just see the old page as-is. this is done in #1626. 

loading blipwise, there's a few flows: 
1. flag not enabled for user, seeing old page -> no blip
2. flag enabled for user, seeing old page -> no blip
3. flag enabled for user, seeing **new** page -> will see old page initially as `grayscale`, will see transition to new page. it's debatable what to show here cos if we block on the whole file fetch, then the user will see **only a loading circle + header** until content is fetched

File: src/components/PageSettingsModal/PageSettingsModal.tsx [hunk 4]
```diff
@@ -11,17 +11,20 @@ import {
   Center,
   Spinner,
   FormControl,
+  Flex,
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/components/PageSettingsModal/PageSettingsModal.tsx:14]: see below response

File: src/layouts/EditPage/EditPage.tsx [hunk 5]
```diff
@@ -14,150 +13,43 @@ import TaskItem from "@tiptap/extension-task-item"
...
-  const getImageSrc = async (src: string) => {
-    const { fileName, imageDirectory } = getImageDetails(src)
-    return mediaService.get({
-      siteName,
-      mediaDirectoryName: imageDirectory || "images",
-      fileName,
-    })
-  }
+  // NOTE: We require this `useEffect` here to sync the variant.
+  // Oddly enough, it's not rerunning when the fetched data changes.
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/EditPage/EditPage.tsx:48]: think you have a conceptual misunderstanding here. `useMemo` returns a memoised value - this is computed **less** than just giving 

```ts
const x = <some_computed_value>
```

as the `x` variable will be recalculated on every render. see [here](https://stackoverflow.com/questions/68513914/react-usememo-without-dependencies-array-vs-empty-array) for more details

File: src/components/PageSettingsModal/ResourcePageSettingsModal.tsx [hunk 6]
```diff
@@ -62,6 +65,7 @@ interface ResourcePageFrontMatter {
   external?: string
   description?: string
   image?: string
+  variant: "tiptap" | "markdown"
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/components/PageSettingsModal/ResourcePageSettingsModal.tsx:68]: sure!

**Pull request merged**